### DISCUSSION TITLE
Parallel programing using pyocd

### DISCUSSION THREAD

**Discussion by Thomas-Lawrence** [pyocd/pyOCD#1125] [Q&A]
Hi,
I am trying to do parallel programming using pyocd. 
Details of my system :
Target : HC32F460
Interface: CMSIS-DAPLink

I was able to flash the target using pyocd and now I am trying to flash several targets in parallel. To start with I connected two targets to two separate interfaces. The pyocd commands are then amended with the unique IDs of these interfaces and created two threads to do flashing, but it fails! I have tested both the interfaces individually and in both cases the target was successfully  flashed.
Is it possible to do parallel programing with pyocd? I am new to python world and any suggestions would help.

**Comment by Hoohaha** [comment]
Are you using the APIs to in your threads?
I am not sure pyocd APIs can support in multiple threads.

I think you can use multiprocessing pool to launch pyocd command line to perform the flashing tasks. It would be easy and more stable.

**Reply by Thomas-Lawrence** [reply]
Hi, am using (subprocess.Popen) inside the thread and 'pyocd list' command works in both threads. Every other command fails.
Please see the log for some of the commands,

**pyocd list**

```
        Probe              Unique ID
---------------------------------------------------------------------------
  0   ARM CMSIS-DAP v1   0000000300190051044ab7bd000007780000000097969921
  1   ARM CMSIS-DAP v1   0000000300190051044ab7bd00000c120000000097969921
```

**pyocd reset in thread 1**
**pyocd reset -t HC32F460xE -M halt -m sw -u 0000000300190051044ab7bd000007780000000097969921 -W**
```
0001024:CRITICAL:__main__:invalid DAP_INFO response length for MAX_PACKET_SIZE
Traceback (most recent call last):
  File "c:\python37\lib\site-packages\pyocd\probe\cmsis_dap_probe.py", line 170, in open
    self._link.open()
  File "c:\python37\lib\site-packages\pyocd\utility\concurrency.py", line 28, in _locking
    return func(self, *args, **kwargs)
  File "c:\python37\lib\site-packages\pyocd\probe\pydapaccess\dap_access_cmsis_dap.py", line 646, in open
    self._packet_size = self._protocol.dap_info(self.ID.MAX_PACKET_SIZE)
  File "c:\python37\lib\site-packages\pyocd\probe\pydapaccess\cmsis_dap_core.py", line 178, in dap_info
    raise DAPAccessIntf.DeviceError("invalid DAP_INFO response length for %s" % id_.name)
pyocd.probe.pydapaccess.dap_access_api.DAPAccessIntf.DeviceError: invalid DAP_INFO response length for MAX_PACKET_SIZE

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "c:\python37\lib\site-packages\pyocd\__main__.py", line 402, in run
    self._COMMANDS[self._args.cmd](self)
  File "c:\python37\lib\site-packages\pyocd\__main__.py", line 645, in do_reset
    session.open(init_board=(not is_hw_reset))
  File "c:\python37\lib\site-packages\pyocd\core\session.py", line 453, in open
    self._probe.open()
  File "c:\python37\lib\site-packages\pyocd\probe\cmsis_dap_probe.py", line 193, in open
    six.raise_from(self._convert_exception(exc), exc)
  File "<string>", line 3, in raise_from
pyocd.core.exceptions.ProbeError: invalid DAP_INFO response length for MAX_PACKET_SIZE
```

**pyocd reset in thread 2**
**pyocd reset -t HC32F460xE -M halt -m sw -u 0000000300190051044ab7bd00000c120000000097969921 -W**
```
0001081:CRITICAL:__main__:invalid DAP_INFO response length for CAPABILITIES
Traceback (most recent call last):
  File "c:\python37\lib\site-packages\pyocd\probe\cmsis_dap_probe.py", line 170, in open
    self._link.open()
  File "c:\python37\lib\site-packages\pyocd\utility\concurrency.py", line 28, in _locking
    return func(self, *args, **kwargs)
  File "c:\python37\lib\site-packages\pyocd\probe\pydapaccess\dap_access_cmsis_dap.py", line 648, in open
    self._capabilities = self._protocol.dap_info(self.ID.CAPABILITIES)
  File "c:\python37\lib\site-packages\pyocd\probe\pydapaccess\cmsis_dap_core.py", line 178, in dap_info
    raise DAPAccessIntf.DeviceError("invalid DAP_INFO response length for %s" % id_.name)
pyocd.probe.pydapaccess.dap_access_api.DAPAccessIntf.DeviceError: invalid DAP_INFO response length for CAPABILITIES

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "c:\python37\lib\site-packages\pyocd\__main__.py", line 402, in run
    self._COMMANDS[self._args.cmd](self)
  File "c:\python37\lib\site-packages\pyocd\__main__.py", line 645, in do_reset
    session.open(init_board=(not is_hw_reset))
  File "c:\python37\lib\site-packages\pyocd\core\session.py", line 453, in open
    self._probe.open()
  File "c:\python37\lib\site-packages\pyocd\probe\cmsis_dap_probe.py", line 193, in open
    six.raise_from(self._convert_exception(exc), exc)
  File "<string>", line 3, in raise_from
pyocd.core.exceptions.ProbeError: invalid DAP_INFO response length for CAPABILITIES
```

**Reply by flit** [reply]
Pyocd is generally thread-safe now when controlling a single probe from multiple threads and within a single process.

It's absolutely possible that there are issues with separate processes, depending on the backend USB driver. Pyocd assumes that the USB driver is safe to use from multiple processes.


@Thomas-Lawrence What OS are you running on, and what version of pyocd?

If Windows, do you have pywinusb or hidapi installed? (pyocd depended on pywinusb for Windows, until version 0.30 when it switched to hidapi, as hidapi has been more stable and continues to be updated whereas pywinusb hasn't been updated in years.) You can check by running `pip list` and looking for both packages. If both are installed, pyocd will select hidapi.

**Reply by Thomas-Lawrence** [reply]
Hi,
I am working on Windows platform. In this case I am trying to use 2 probes each connected with a target controller.

Regarding the driver, I followed the libusb installation documentation and was working fine for a single probe but it started giving issues when I tried to control two probes from separate threads at the same time.

Yes I have hidapi installed, see below

Package                Version
---------------------- ------------
cmsis-pack-manager     0.2.10
hidapi                 0.10.0.post1
pyocd                  0.30.2
pypiwin32              223
pyusb                  1.1.0
pywin32                228

**Reply by flit** [reply]
Thanks for the info.

It would be helpful to know which version of CMSIS-DAP is being used. (DAPLink supports both.) That changes which USB backend is used in pyocd. libusb will only be used for CMSIS-DAPv2 (and STLink and picoprobe), which use WinUSB instead of HID since it's faster while still not requiring a kernel driver on Windows.

Would you mind running:

    pyocd cmd -u 0000000300190051044ab7bd00000c120000000097969921

then when the "pyocd>" prompt appears run this (the "$" is part of the command and must be typed; it starts a Python expression):

    $probe._probe._link._interface

This will print something like:

    <pyocd.probe.pydapaccess.interface.hidapi_backend.HidApiUSB object at 0x10b575070>

**Reply by Thomas-Lawrence** [reply]
Hi,
CMSIS DAP version used is 'ARM CMSIS-DAP v1'.
Please see log below,

X:\>**pyocd cmd -u 0000000300190051044ab7bd000007780000000097969921 -t HC32F460xE**
Connected to HC32F460xE [Lockup]: 0000000300190051044ab7bd000007780000000097969921

pyocd> **$probe._probe._link._interface**
<pyocd.probe.pydapaccess.interface.hidapi_backend.HidApiUSB object at 0x00000193F967B288>

pyocd> **exit**

**Reply by flit** [reply]
Perfect, thanks! That confirms that pyOCD is using hidapi. Let's try an experiment by switching to pywinusb, and see if that resolves your problem. Even if it does fix things, there's still a need for a real fix, either in pyocd or upstream in hidapi. We can't just switch back to pywinusb, since there a some debug probes that don't work with it and it hasn't been updated in years.

Please run these commands:

```console
pip install pywinusb
set PYOCD_USB_BACKEND=pywinusb
```

Then try your tests with pyocd again. Setting the `PYOCD_USB_BACKEND` environment variable overrides the default selection.

I'll also try to reproduce in my Windows VM and work on fixing the problem.

**Reply by Thomas-Lawrence** [reply]
Hi Flit,
Finally the parallel programing works :). 
It was actually my mistake, one of the commands was actually pointing to the same device from both threads , which caused issues with the DAP response. I would like to thankyou for your support.

From my testing both hidapiusb and pywinusb backends works fine for parallel programing.
